### DISCUSSION TITLE
Send data to SED devices

### DISCUSSION THREAD

**Discussion by ksathara** [openthread/openthread#9351] [Q&A]
Hi,
How can we send data to SED devices? I read about 'Indirect communication method', but couldnt find out how to implement it.
I am using a CoAP client/Server architecture, where the CoAP clients are SEDs.

Thanks for you help,
Kaushalya

**Comment by Irving-cl** [comment]
Hi @ksathara,

When a router tries to send some data to a SED, it will first 'buffer' this message into a queue for this SED. And then, the Router waits for a Mac Data Request sent by the SED. After the Router receives the Mac Data Request, it will send that message to the SED. An SED will send a Mac Data Request periodically. And the period can be set by [`pollperiod` command](https://github.com/openthread/openthread/tree/main/src/cli#pollperiod-pollperiod).

Back to the question, if both your client and server are OT devices or standard Thread devices, you don't need to implement anything. Just make sure your SED is configured correctly and sends Mac Data Request periodically.

**Comment by ksathara** [comment]
Hi Li,

Many thanks for your prompt reply. In case the client was out of range at
the time the host sent the data, does the router keep the packet
indefinitely in its queue until the client comes back?

Cheers,
Kaushalya

**Reply by abtink** [reply]
If a parent does not receive a message (data poll) from a sleepy child within the child's specified timeout interval, the parent will remove the child from the child table and clear any pending messages for the child. 

The child's timeout interval is communicated to the parent when the child first attaches, and is typically set to 240 seconds, but can be changed by the child (please see `otThreadSetChildTimeout()` API or [`childtimeout`](https://github.com/openthread/openthread/tree/main/src/cli#childtimeout) CLI command).

**Comment by ksathara** [comment]
Hi Abtin,

Thanks. I have two questions.
1. What APIs can I use to send data to SED client? I use Coap client/server
architecture. I can see in the CoAP client, it uses coap_send_request ()

**Reply by abtink** [reply]
Thread is an IPv6 network. The fact that the receiver is a Sleepy End Device (SED) does not affect how the APIs are used on the sender. The mechanism for a parent to queue messages destined for its sleepy chil (i.e., indirect transmission) is handled entirely by the Thread stack itself. 

OpenThread provides different levels of APIs:
- You can check [`ip6.h`](https://github.com/openthread/openthread/blob/main/include/openthread/ip6.h) APIs, in particular `otIp6Send()` to send a prepared IPv6 message.
- Of if you want to use UDP sockets, check [`udp.h`](https://github.com/openthread/openthread/blob/main/include/openthread/udp.h) APIs (which define UDP socket and APIs to send and receive using the socket)
- Of if you want to use COAP protocol check the [`coap.h`](https://github.com/openthread/openthread/blob/main/include/openthread/coap.h) APIs.

**Comment by ksathara** [comment]
Hi Abtin,

Thanks. I have a couple of questions.
1. What APIs can I use to send data to SED clients? I use Coap
client/server architecture. I can see in the CoAP client, it uses
coap_send_request
() function. Can I use the same from CoAP server side as well?
2. When a Client is out of range for a long time, it takes a long time to
connect back. I can see the retry time is extended in each failed reconnect
attempt or something like that. How can this be controlled, for ex. send a
poll signal by press of a switch?

Thanks,
Kaushalya

**Reply by abtink** [reply]
Question 1 ) I already answered -> https://github.com/openthread/openthread/discussions/9351#discussioncomment-6697402

Question 2)  I think you are talking about "attch backoff" feature.
Please read the config and their documentaion to learn more about this and how to configure it-> https://github.com/openthread/openthread/blob/main/src/core/config/mle.h#L113

You can reset the attach back off process and trigger an immediate attach attempt on SED by using `otThreadSetEnabled()` API , i.e., disable Thread operation and enable it back immediately.

**Comment by ksathara** [comment]
Hi Abtin,

What API can I use to retrieve the IP address of a client? I know the
extended address, and EUI64 of the client.

Thanks,
Kaushalya

**Reply by abtink** [reply]
IP address discovery is generally done through service registration and DNS look up.
You can check https://openthread.io/codelabs/openthread-border-router#1 as an example of this.

Depending on your use-case you can do other thing. For example, on parent node you can get the addresses of MTD/SED children. Check [`otThreadGetChildNextIp6Address`](https://github.com/openthread/openthread/blob/70adcabca6270eead21f126533c697d4d1637b77/include/openthread/thread_ftd.h#L577-L597).  Or its equivalent cli command [`childip`](https://github.com/openthread/openthread/blob/main/src/cli/README.md#childip)

**Comment by ksathara** [comment]
Hi,

I found the function 'otThreadGetRloc ()' which return the complete IPV6 of
the shield nodes except the last two octets, which is the rloc of the
child. So I can find the IPV6 of a child node by getting an IPV6 by
otThreadGetRloc
() and adding the RLOC of the particular child node to replace the last two
octets. Is this correct?

Cheers,
Kaushalya

**Comment by ksathara** [comment]
Hi Abtin,

otThreadGetChildNextIp6Address() could only be used when the children 
are directly connected to the leader/router, isnt it? In case a child is 
connected via a router this would not work as I cant get a index for 
that child.

If so, how can it be resolved?

Cheers,
Kaushalya


------ Original Message ------
From "Abtin Keshavarzian" ***@***.***>
To "openthread/openthread" ***@***.***>
Cc "Kaushalya Satharasinghe" ***@***.***>; "Mention" 
***@***.***>
Date 8/12/2023 2:01:24 AM
Subject Re: [openthread/openthread] Send data to SED devices (Discussion 
#9351)

**Reply by jwhui** [reply]
I suggest leveraging DNS-based Service Discovery. See the [Thread Border Router - Bidirectional IPv6 Connectivity and DNS-Based Service Discovery](https://openthread.io/codelabs/openthread-border-router) codelab for more information.

**Comment by ksathara** [comment]
Hi,

I have extracted the client IP6 from 'message_info->mPeerAddr'. I have
tried using 'coap_send_request ()' using this IP6 address but I get
'Transmission failed: 9'. I tried searching what the error code 9 says, but
couldn't find out.
I do not have implemented any on the client side to receive this. Is that
could be the reason?

What APIs can I use to send a data packet to a CoAP client?
How to setup a CoAP client to receive data?

Cheers,
Kaushalya
